File: android/src/main/kotlin/com/example/trusted_time/TrustedTimePlugin.kt
package com.example.trusted_time

import android.os.SystemClock
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import android.os.Handler
import android.os.Looper

/**
 * TrustedTimePlugin - Native Android implementation for TrustedTime Flutter plugin.
 *
 * This plugin serves as the low-level bridge to the Android hardware clocks. 
 * While the wall clock (System.currentTimeMillis()) can be easily manipulated by users,
 * this plugin leverages the monotonic 'Elapsed Realtime' clock which is hardware-backed 
 * and independent of user-controlled time settings.
 *
 * Key Features:
 * - Monotonic Uptime: Provides access to time since boot, including deep sleep.
 * - Broadcast Monitoring: Listens for system events that invalidate the time trust.
 * - Reboot Resilience: Helps the Dart layer detect when hardware counters have reset.
 */
class TrustedTimePlugin : FlutterPlugin, MethodCallHandler {
    
    /// The MethodChannel that will handle communication between Flutter and native Android
    private var channel: MethodChannel? = null

    companion object {
        /// Static reference to the channel for use by broadcast receivers
        @Volatile
        private var staticChannel: MethodChannel? = null

        /**
         * Notifies Flutter that the system clock has been tampered with.
         * Called by ClockChangeReceiver and BootReceiver.
         */
        fun notifyClockTampered() {
            Handler(Looper.getMainLooper()).post {
                try {
                    staticChannel?.invokeMethod("onClockTampered", null)
                } catch (e: Exception) {
                    // Channel might not be ready or engine is detaching
                    // Silently fail to avoid crashes
                }
            }
        }
    }

    /**
     * Called when the plugin is attached to the Flutter engine.
     * Sets up the MethodChannel for communication.
     */
    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "trusted_time")
        channel?.setMethodCallHandler(this)
        staticChannel = channel
    }

    /**
     * Handles method calls from Flutter.
     *
     * Supported methods:
     * - getPlatformVersion: Returns Android version string
     * - getUptimeMs: Returns monotonic uptime in milliseconds (includes deep sleep)
     */
    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }
            "getUptimeMs" -> {
                // We use elapsedRealtime() instead of uptimeMillis().
                // uptimeMillis() stops when the CPU enters deep sleep, which would 
                // introduce drift. elapsedRealtime() continues counting in sleep modes,
                // providing the most reliable monotonic reference available on Android.
                val uptimeMs = SystemClock.elapsedRealtime()
                result.success(uptimeMs)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    /**
     * Called when the plugin is detached from the Flutter engine.
     * Cleans up the MethodChannel.
     */
    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel?.setMethodCallHandler(null)
        channel = null
        staticChannel = null
    }
}

File: ios/trusted_time/Sources/trusted_time/TrustedTimePlugin.swift
import Flutter
import UIKit

/**
 * TrustedTimePlugin - Native iOS implementation for TrustedTime.
 *
 * This plugin provides the hardware-backed monotonic clock reference needed 
 * to defy system wall-clock manipulation. On iOS, we leverage `ProcessInfo.systemUptime`
 * which is a highly accurate hardware counter that is unaffected by user changes
 * to the Date & Time settings.
 *
 * Design Note: Unlike Android, iOS restricts background broadcast reception for 
 * many time-related events. We mitigate this by observing `NSSystemClockDidChange`
 * while the app is active and relying on uptime-delta validation in the Dart layer.
 */
public class TrustedTimePlugin: NSObject, FlutterPlugin {
    
    /// The FlutterMethodChannel for communication with Dart
    private static var channel: FlutterMethodChannel?
    
    /**
     * Registers the plugin with the Flutter engine.
     * Called automatically by Flutter when the plugin is loaded.
     *
     * - Parameter registrar: The plugin registrar provided by Flutter
     */
    public static func register(with registrar: FlutterPluginRegistrar) {
        channel = FlutterMethodChannel(
            name: "trusted_time",
            binaryMessenger: registrar.messenger()
        )
        
        let instance = TrustedTimePlugin()
        registrar.addMethodCallDelegate(instance, channel: channel!)
        
        // Start observing system clock changes
        ClockObserver.start(channel: channel!)
    }
    
    /**
     * Handles method calls from Flutter.
     *
     * Supported methods:
     * - getPlatformVersion: Returns iOS version string
     * - getUptimeMs: Returns monotonic uptime in milliseconds
     *
     * - Parameters:
     *   - call: The method call from Flutter
     *   - result: The result callback to send data back to Flutter
     */
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)
            
        case "getUptimeMs":
            // systemUptime is the most reliable monotonic clock on iOS.
            // It represents seconds since the last kernel boot.
            // Converting to Int64 milliseconds for cross-platform parity with Android.
            let uptimeSeconds = ProcessInfo.processInfo.systemUptime
            let uptimeMs = Int64(uptimeSeconds * 1000)
            result(uptimeMs)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}
